src/controllers/subscription.js

```diff
@@ -24,114 +30,178 @@ const getSubscriptionsByID = (req, res) => {
+
+    if (!existingClass || !existingMember) {
+      return res.status(404).json({
+        error: true,
+        message: 'Class or member not found',
+      });
+    }
+
+    const existingSubscription = await Subscription.findOne({ classes, member, date });
+
+    if (existingSubscription) {
+      return res.status(400).json({
+        error: true,
+        message: 'Subscription already exists',
+        data: existingSubscription,
+      });
+    }
+
+    const newSubscription = await Subscription.create({
+      classes,
+      member,
+      date,
+    });
+
+    return res.status(201).json({
       error: false,
       message: 'Subscription created successfully',
       data: newSubscription,
-    }))
-    .catch((error) => res.status(500).json({
+    });
+  } catch (error) {
+    return res.status(500).json({
       error: true,
-      message: error,
+      message: error.message,
       data: undefined,
-    }));
+    });
+  }
 };
 
-const updateSubscription = (req, res) => {
+const updateSubscription = async (req, res) => {
   const { id } = req.params;
   const { classes, member, date } = req.body;
 
-  if (!mongoose.Types.ObjectId.isValid(id)) {
+  if (!mongoose.isValidObjectId(id)
+  || (classes && !mongoose.isValidObjectId(classes))
+  || (member && !mongoose.isValidObjectId(member))) {
     return res.status(400).json({
       error: true,
-      message: 'ID is not valid',
+      message: 'ID or class or member ID is not valid',
     });
   }
-  return Subscription.findByIdAndUpdate(
-    id,
-    {
-      classes,
-      member,
-      date,
-    },
-    { new: true },
-  )
-    .then((subscriptions) => {
-      if (!subscriptions) {
-        return res.status(404).json({
-          error: true,
-          message: 'Subscription not found',
-          data: undefined,
-        });
-      }
-      return res.status(200).json({
-        error: false,
-        message: 'Subscription updated successfully',
-        data: subscriptions,
+
+  try {
+    const existingClass = await Class.findById(classes);
+    const existingMember = await Member.findById(member);
+
+    if ((classes && !existingClass) || (member && !existingMember)) {
+      return res.status(404).json({
+        error: true,
+        message: 'Class or member not found',
       });
-    })
-    .catch((error) => res.status(500).json({
+    }
+
+    const existingSubscription = await Subscription.findOne({ classes, member, date });
+
+    if (existingSubscription && existingSubscription.id.toString() !== id) {
+      return res.status(400).json({
+        error: true,
+        message: 'Subscription already exists',
+        data: existingSubscription,
+      });
+    }
+
+    const updatedSubscription = await Subscription.findByIdAndUpdate(
+      id,
+      {
+        classes,
+        member,
+        date,
+      },
+      { new: true },
+    ).populate('classes member');
+
+    if (!updatedSubscription) {
```

Comment:
You could check if the subscription with the id is found at the beginning of the controller by using the findById method, and then you run the rest of the validations, wdyt?